### PR TITLE
Add CVE-2024-8963

### PR DESCRIPTION
**Pull request by nmasdoufi-ol**
### Summary
This PR implements a detection script for CVE-2024-8963, a path traversal vulnerability in Ivanti Cloud Services Appliance (CSA) versions before 4.6 Patch 519. The vulnerability allows remote unauthenticated attackers to access restricted functionality. We've also added comprehensive unit tests to ensure the reliability and accuracy of the detection script.

### Changes
- Implemented `CVE20248963Exploit` class in `agent/exploits/cve_2024_8963.py`
- Added unit tests in `tests/exploits/test_cve_2024_8963.py`

### Implementation Details
The detection script does the following:
1. Identifies Ivanti CSA instances by checking for the specific title in the HTML.
2. Scans the target by requesting the `/allowed/ivanti-logo.png` file and checking the `Last-Modified` header.
3. Compares the `Last-Modified` date with the patch release date (September 10, 2024).
4. Reports the target as vulnerable if the `Last-Modified` date is before the patch date.

### Testing
The unit tests cover the following scenarios:
- Detecting a vulnerable Ivanti CSA instance
- Correctly identifying a patched (safe) Ivanti CSA instance
- Handling missing `Last-Modified` headers
- Dealing with invalid date formats in the `Last-Modified` header
- Rejecting non-Ivanti CSA targets
- Gracefully handling request exceptions

It was also tested against a real target and it did reported a vulnerability as it was expected:

image

### CODE REVIEW

File: agent/exploits/cve_2024_8963.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+"""Agent Asteroid implementation for CVE-2024-8963"""
+
+import re
+import logging
+import datetime
+import requests
+
+from urllib import parse as urlparse
```
Comment (elyousfi5) [inline agent/exploits/cve_2024_8963.py:8]:
```suggestion

import requests
from urllib import parse as urlparse
```

File: agent/exploits/cve_2024_8963.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+RISK_RATING = "CRITICAL"
+PATCH_DATE = datetime.datetime(2024, 9, 10)
+
+MAX_REDIRECTS = 2
+DEFAULT_TIMEOUT = 90
+
+
+@exploits_registry.register
+class CVE20248963Exploit(webexploit.WebExploit):
+    accept_request = definitions.Request(method="GET", path="/")
```
[the remark this answers is not in the document]
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_8963.py:27]: The logic is implemented in the accept_pattern

[the remark this answers is not in the document]
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_8963.py:27]: Also `accept_request` is used within the `WebExploit` class

File: agent/exploits/cve_2024_8963.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            req = requests.Request(
+                method=self.check_request.method,
+                url=target_endpoint,
+            ).prepare()
+            resp = session.send(req, timeout=DEFAULT_TIMEOUT)
+        except requests.RequestException as e:
+            logging.error("HTTP Request failed: %s", e)
+            return vulnerabilities
+
+        if "Last-Modified" not in resp.headers:
```
Comment (Zakaria-Latif) [inline agent/exploits/cve_2024_8963.py:66]: Consider using  constants for reusability:
HEADER_LAST_MODIFIED = "Last-Modified"

**Review by PiranhaSa — APPROVED** [review]

File: agent/exploits/cve_2024_8963.py (hunk 3 above)
Comment (elyousfi5) [reply agent/exploits/cve_2024_8963.py:66]: +1

File: agent/exploits/cve_2024_8963.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from agent.exploits import webexploit
+from agent import exploits_registry
+from agent import definitions
+
+
+VULNERABILITY_TITLE = "Path Traversal in Ivanti CSA"
+VULNERABILITY_REFERENCE = "CVE-2024-8963"
+VULNERABILITY_DESCRIPTION = """Path Traversal in the Ivanti CSA before 4.6 Patch 519 allows a remote unauthenticated attacker to access restricted functionality."""
+RISK_RATING = "CRITICAL"
+PATCH_DATE = datetime.datetime(2024, 9, 10)
```
Comment (3asm) [inline agent/exploits/cve_2024_8963.py:19]: Please document where this value came from.

**Review by ybadaoui-ostorlab — APPROVED** [review]

File: agent/exploits/cve_2024_8963.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            )
+        except ValueError as e:
+            logging.error("Couldn't parse date string and format: %s", e)
+            return vulnerabilities
+
+        if last_modified < PATCH_DATE:
+            vulnerability = self._create_vulnerability(target)
+            vulnerabilities.append(vulnerability)
+
+        return vulnerabilities
```
Comment (3asm) [inline agent/exploits/cve_2024_8963.py:81]:
```suggestion
    def check(self, target: definitions.Target) -> list[definitions.Vulnerability]:
        """Rule to detect specific vulnerability on a specific target.

        Args:
            target: Target to scan

        Returns:
            List of identified vulnerabilities.
        """
        session = requests.Session()
        session.max_redirects = MAX_REDIRECTS
        session.verify = False

        target_endpoint = urlparse.urljoin(target.origin, self.check_request.path)

        try:
            req = requests.Request(
                method=self.check_request.method,
                url=target_endpoint,
            ).prepare()
            resp = session.send(req, timeout=DEFAULT_TIMEOUT)
        except requests.RequestException as e:
            logging.error("HTTP Request failed: %s", e)
            return []

        if "Last-Modified" not in resp.headers:
            return []

        try:
            last_modified = datetime.datetime.strptime(
                resp.headers["Last-Modified"], "%a, %d %b %Y %H:%M:%S %Z"
            )
        except ValueError as e:
            logging.error("Couldn't parse date string and format: %s", e)
            return []

        if last_modified < PATCH_DATE:
            vulnerability = self._create_vulnerability(target)
            return [vulnerability]
        else:
            return []
```
Much clearer intent.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/exploits/cve_2024_8963.py (hunk 3 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_8963.py:64]: It's the unit test when that header not in the response that's missing. I covered when it's present.

File: agent/exploits/cve_2024_8963.py (hunk 4 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_8963.py:23]: Done.

File: agent/exploits/cve_2024_8963.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import requests
+from urllib import parse as urlparse
+
+from agent.exploits import webexploit
+from agent import exploits_registry
+from agent import definitions
+
+
+VULNERABILITY_TITLE = "Path Traversal in Ivanti CSA"
```
Comment (benyissa) [inline agent/exploits/cve_2024_8963.py:15]:
```suggestion
VULNERABILITY_TITLE = "Path Traversal in Ivanti CSA before 4.6 Patch 519"
```
I would make the title more specific.

Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_8963.py:15]: I wouldn't.

**Review by 3asm — APPROVED** [review]

**Pull request merged**